src/crimson/net/SocketConnection.cc

```diff
@@ -65,38 +64,100 @@ SocketConnection::get_messenger() const {
 seastar::future<bool> SocketConnection::is_connected()
 {
   return seastar::smp::submit_to(shard_id(), [this] {
-      return !send_ready.failed();
+      return write_state == write_state_t::open;
     });
 }
 
-seastar::future<> SocketConnection::send(MessageRef msg)
+void SocketConnection::write_event()
 {
-  return seastar::smp::submit_to(shard_id(), [this, msg=std::move(msg)] {
-      if (state == state_t::closing)
-        return seastar::now();
-      return seastar::with_gate(pending_dispatch, [this, msg=std::move(msg)] {
-          return do_send(std::move(msg))
-            .handle_exception([this] (std::exception_ptr eptr) {
-              logger().warn("{} send fault: {}", *this, eptr);
+  if (write_dispatching) {
+    // already dispatching
+    return;
+  }
+  write_dispatching = true;
+  switch (write_state) {
+   case write_state_t::open:
+   case write_state_t::delay:
+    seastar::with_gate(pending_dispatch, [this] {
+      return seastar::repeat([this] {
```

Comment:
`seastar::repeat()` here acts as a do-while loop trying to consume `_keepalive` and `_keepalive_ack` events and deliver all the messages in the `out_q`. That's why we always schedule a `write_event()` after we enqueue a message or attempt to send a keepalive or keepalive-ack.

The only "condition" here is the `state_changed` promise, which is to promise waiters (currently only `write_event()`) that current state shall be changed in the future, and notify them when it happens. That's why when a write_state is changed, it fulfills the promise of the previous `state_changed` to notify waiters about this new state, and set a new promise to allow the latter waiters wait for the current state to be changed... Do you think it is better to be a `seastar::condition_variable` instead of a `promise`?

Reply:
@cyx1231st i have a strong feeling that you are replicate an "epoll" using `state_changed` promise. and there could be multiple waiters waiting for the `state_changed`, and not *all* changes are expected, if the waiter is not satisfied with the new state, it will reschedule itself and wait for another `state_changed`. the reason i prefer `condition_variable` is that it allows you to specify a predicate. it saves the efforts in the application level to ignore the unexpected changes.

but, once a connection is established, how can this batch processing save the i/o overhead?

Reply:
`seastar::condition_variable` is more complex than `promise`, it wraps `seastar::semaphore`, and further wraps `promise`.

From my existing understandings, only `write_event()` will wait for the `state_changed`, and there is at most one pending `write_event()` at the same time because of the `write_dispatching`. And as for current simple design, it doesn't matter who check the changes, I think.

> but, once a connection is established, how can this batch processing save the i/o overhead?

* The `out_q` implements SEDA in seastar, so `conn->send(msg)` can be done very fast, and messages are enqueued and batched.
* The batching saves the `socket->flush()` effort, `write_event()` will flush the buffer only when the `out_q` is truly empty, or `write_state` becomes `write_state_t::delay` (will add it later). It allows seastar to batch its internal sending buffer.
* The `write_event()` with `repeat()` loop is faster than chaining futures for each message.
* We can further improve the performance by combining the buffer of the batched messages, and send them on the wire in one continuation.

Reply:
have you considered using `seastar::execution_stage` for batching the write? namely, to pump messages out of `out_q`, and to flush the socket after the last message is written to socket.